### PR TITLE
Fix failure to load System.IO.Compression.ZipFile due to automatic binding redirect

### LINKED ISSUE #2292: Unable to load System.IO.Compression.ZipFile when targeting 4.7.1 after the new conflict resolution changes.
It was reported internally that after the new changes that went in for conflict resolution, apps are running into an issue where they can't find System.IO.Compression.ZipFile v.4.0.3.0 at runtime. The issue seems to be that the new logic will add a Reference to the platform ZipFile assembly, which will cause a conflict with the package reference and RAR will suggest a binding redirect to the 4.0.3.0 version. The problem is that if that binding redirect is added, is that System.IO.Compression.ZipFile is not part of the support package, so we won't be able to find an assembly that matches that version inbox(because the inbox version is 4.0.0.0). There are a few ways of working around this, one would be to manually add a binding redirect to your project that binds down to the platform version, that way assembly will always be able to load.
```xml
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.IO.Compression.ZipFile" publicKeyToken="b77a5c561934e089" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
```

The other option, is to add System.IO.Compression.ZipFile to the support package version 4.0.3.0 and that way the binding redirect that gets added automatically will just work fine.
cc: @dsplaisted

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #2292

There is a bug with the unification table for .NET 4.7.1, and the [workaround](https://github.com/Microsoft/dotnet/blob/master/releases/net471/KnownIssues/623552-BCL%20Higher%20assembly%20versions%20that%204.0.0.0%20for%20System.IO.Compression.ZipFile%20cannot%20be%20loaded%20without%20a%20binding%20redirect.md) is to add the following binding redirect:

```xml
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.IO.Compression.ZipFile" publicKeyToken="b77a5c561934e089" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
```

With the fix for #2221, this is getting overridden with an automatic binding redirect to version 4.0.3.0, which fails at runtime because it disables the unification table, and because the version of the runtime file is 4.0.0.0.

Since the suggested binding redirect is wrong in this case, this PR removes that suggested redirect after RAR suggests it.

### CODE REVIEW

**Comment by dsplaisted** [comment]
@joperezr @AlexGhiondea @nguerrera for review

**Review by joperezr — CHANGES_REQUESTED** [review]
The changes to remove the .dll LGTM, but if @AlexGhiondea confirms that another valid solution is to ask people to remove the redirect, I much rather do that as oposed to having your custom `RemoveZipFileSuggestedRedirect` target.

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT Release

**Comment by dsplaisted** [comment]
@ericstj for review

File: src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets [hunk 1]
```diff
@@ -135,4 +135,15 @@ Copyright (c) .NET Foundation. All rights reserved.
       </ReferenceCopyLocalPaths>
     </ItemGroup>
   </Target>
+
+  <Target Name="RemoveZipFileSuggestedRedirect"
+          BeforeTargets="GenerateBindingRedirects"
+          DependsOnTargets="ResolveAssemblyReferences"
+          Condition="'$(_TargetFrameworkVersionWithoutV)' &gt;= '4.7.1'">
```
[the remark this answers is not in the document]
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets]: Per our discussion, I don't think that it is dangerous to remove the binding redirect for all frameworks above 4.7.1, as those frameworks should always have the assembly in the unification table, and thus the binding redirect should never be needed (nor correct).

**Comment by ericstj** [comment]
Why are you adding a code fix for a bad side-effects of a workaround?  Just update the workaround so it accounts for the problem.

If you want to make a code-fix then the right fix is to make RAR more aware of the targeting pack's file list and not suggest redirects for things that are in the framework.

**Comment by dsplaisted** [comment]
@dotnet-bot test Ubuntu14.04 Release

File: src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets [hunk 2]
```diff
@@ -135,4 +135,15 @@ Copyright (c) .NET Foundation. All rights reserved.
       </ReferenceCopyLocalPaths>
     </ItemGroup>
   </Target>
+
+  <Target Name="RemoveZipFileSuggestedRedirect"
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets]: Make this "private" with leading "_"

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets]: Please comment this. I know some poor souls are going to read this and wonder what it's about. This is bizarre enough that tests aren't enough convey "why" we're doing it.

**Review by nguerrera — APPROVED** [review]
Very reluctantly signing off with minor comments. The special casing to workaround a workaround feels wrong, but after must discussion elsewhere, I agree it's the least bad option here.

**Comment by dsplaisted** [comment]
@dotnet-bot test this please

**Comment by dsplaisted** [comment]
@dotnet-bot test this please

**Comment by dsplaisted** [comment]
@dotnet-bot test this please

**Pull request merged**